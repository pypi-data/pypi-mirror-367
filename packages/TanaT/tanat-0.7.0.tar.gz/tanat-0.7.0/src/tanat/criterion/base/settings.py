#!/usr/bin/env python3
"""
Base class for criterion.
"""

import dataclasses

from .enum import CriterionLevel


@dataclasses.dataclass
class Criterion:
    """
    Base class for criterion.
    """

    _REGISTER_NAME = None

    def get_compatibility_levels(self, max_level=CriterionLevel.TRAJECTORY):
        """List the compatibility levels of the criterion.
        Compatibility levels are marked with decorators.

        Args:
            max_level (CriterionLevel, optional): The maximum level to consider.
                Defaults to Trajectory. Entity < Sequence < Trajectory

        Returns a list of CriterionLevel representing the compatibility levels of the
        criterion. The levels are "sequence","entity" or "trajectory".
        """
        if isinstance(max_level, str):
            max_level = CriterionLevel.from_str(max_level)

        compatibility_levels = getattr(self, "COMPATIBILITY_LEVELS", set())
        return [level for level in compatibility_levels if level <= max_level]
