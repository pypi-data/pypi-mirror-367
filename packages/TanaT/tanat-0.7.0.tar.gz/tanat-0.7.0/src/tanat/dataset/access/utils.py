#!/usr/bin/env python3
"""
User data access utils.
"""

from .zenodo import ZenodoAccessor


def access(data_type, cache_dir=None, force=False):
    """
    Access a dataset from Zenodo and return a ready-to-use object.

    Depending on the type, this may return:
    - a DataFrame (for file-based datasets: CSV, Parquet, etc.)
    - a database connection (for SQL-based datasets)

    Data is cached locally after the first download unless `force=True` is set.

    Args:
        data_type (str): Name of the dataset registered in ZenodoAccessor.
        cache_dir (str, optional): Directory used for caching. Defaults to system temp directory.
        force (bool): If True, forces re-download even if data is already cached.

    Returns:
        A usable object for interacting with the dataset.

    Raises:
        ValueError: If `data_type` is not registered in the accessor.

    Examples:
        >>> # Access a MVAD CSV dataset as a DataFrame
        >>> df = access("mvad")

        >>> # Access a SQL database as a connection
        >>> conn = access("mimic4")
    """
    accessor = ZenodoAccessor.init(data_type, cache_dir=cache_dir)
    return accessor.get(force=force)
