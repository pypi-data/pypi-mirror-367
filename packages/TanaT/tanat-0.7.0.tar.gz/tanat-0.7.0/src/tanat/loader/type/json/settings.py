#!/usr/bin/env python3
"""
Loader for JSON files.
"""

from pydantic.dataclasses import dataclass
from pypassist.dataclass.decorators.viewer.decorator import viewer
from pypassist.fallback.typing import Dict


@viewer
@dataclass
class JSONLoaderSettings:
    """
    Settings for JSON loader

    Attributes:
        filepath: Path to the JSON file to load
        encoding: File encoding (utf-8 by default)
        json_load_kwargs: Additional arguments to pass to json.load
    """

    filepath: str
    encoding: str = "utf-8"
    json_load_kwargs: Dict = None

    def __post_init__(self):
        self.json_load_kwargs = self.json_load_kwargs or {}
