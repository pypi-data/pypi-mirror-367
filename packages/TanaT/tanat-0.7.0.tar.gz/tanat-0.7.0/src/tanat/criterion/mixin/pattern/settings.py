#!/usr/bin/env python3
"""
Pattern criterion.
"""

from typing import Union

from pydantic.dataclasses import dataclass
from pypassist.dataclass.decorators.viewer.decorator import viewer
from pypassist.fallback.typing import Dict, List

from ...base.settings import Criterion
from ...decorator.entity import entity_compatible
from ...decorator.sequence import sequence_compatible
from .enum import PatternOperator


@viewer
@entity_compatible
@sequence_compatible
@dataclass
class PatternCriterion(Criterion):
    """
    Flexible pattern-based filtering for sequences and entities.

    Enables complex pattern matching across data columns with advanced
    configuration options.

    Args:
        pattern (Dict[str, Union[str, List[str]]]):
            Column patterns to match. Supports:
            - Plain strings
            - Lists of strings
            - Regex patterns (prefix with 'regex:')

        contains (bool, optional):
            - True: Include sequences matching pattern
            - False: Exclude sequences matching pattern
            Defaults to True.

        case_sensitive (bool, optional):
            Whether matching is case-sensitive. Defaults to True.

        operator (PatternOperator, optional):
            Logical operator between patterns:
            - AND: All patterns must match
            - OR: At least one pattern matches
            Defaults to AND.

    Example:
        # Find sequences with emergency events
        PatternCriterion(pattern={'event_type': 'EMERGENCY'})
    """

    pattern: Dict[str, Union[str, List[str]]]
    contains: bool = True
    case_sensitive: bool = True
    operator: PatternOperator = PatternOperator.AND

    _REGISTER_NAME = "pattern"
