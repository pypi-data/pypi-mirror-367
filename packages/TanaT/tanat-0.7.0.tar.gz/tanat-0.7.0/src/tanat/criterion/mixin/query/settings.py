#!/usr/bin/env python3
"""
Settings for query criterion.
"""

from pydantic.dataclasses import dataclass
from pypassist.dataclass.decorators.viewer.decorator import viewer

from ...base.settings import Criterion
from ...decorator.entity import entity_compatible
from ...decorator.sequence import sequence_compatible


@viewer
@entity_compatible
@sequence_compatible
@dataclass
class QueryCriterion(Criterion):
    """
    Pandas-based filtering criterion for sequences and entities.

    Allows complex data filtering using pandas query syntax. Compatible with
    sequence and entity levels, enabling flexible data selection based on
    column conditions.

    Args:
        query (str): Pandas-style query string to filter data.
            Supports complex boolean conditions across columns.

    Example:
        # Filter sequences where age > 50 and gender is 'M'
        criterion = QueryCriterion(query="age > 50 and gender == 'M'")
    """

    query: str

    _REGISTER_NAME = "query"
