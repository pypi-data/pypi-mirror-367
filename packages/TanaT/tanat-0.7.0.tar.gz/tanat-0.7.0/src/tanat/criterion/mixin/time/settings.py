#!/usr/bin/env python3
"""
Settings for time-based filtering.
"""

from typing import Optional, Union
from datetime import datetime

from pydantic.dataclasses import dataclass
from pypassist.dataclass.decorators.viewer.decorator import viewer

from ...base.settings import Criterion
from ...decorator.entity import entity_compatible
from ...decorator.sequence import sequence_compatible


@viewer
@entity_compatible
@sequence_compatible
@dataclass
class TimeCriterion(Criterion):
    """
    Flexible time-based filtering for sequences and entities.

    Enables precise temporal filtering with multiple time constraints and
    different containment modes depending on the filtering level.

    Args:
        start_before (datetime, optional):
            Filter sequences or entities starting before this timestamp.
        start_after (datetime, optional):
            Filter sequences or entities starting after this timestamp.
        end_before (datetime, optional):
            Filter sequences or entities ending before this timestamp.
        end_after (datetime, optional):
            Filter sequences or entities ending after this timestamp.
        duration_within (bool, optional):
            For entity-level filtering on state/interval entities:
            - True: Entity must be entirely contained within the time range
            - False: Entity can partially overlap with the time range
            Defaults to False. Has no effect on event entities.
        sequence_within (bool, optional):
            For sequence-level filtering (event, interval, state):
            - True: Entire sequence must be contained within the time range
            - False: Sequence can partially overlap with the time range
            Defaults to False.
        date_format (str, optional):
            Custom datetime parsing format for string inputs.

    Examples:
        # Filter entities with partial overlap (default behavior)
        TimeCriterion(
            start_after=datetime.now(),
            end_before=datetime.now() + timedelta(days=90)
        )

        # Only keep state/interval entities fully contained in time range
        TimeCriterion(
            start_after=datetime(2024, 1, 1),
            end_before=datetime(2024, 12, 31),
            duration_within=True
        )

        # Only keep sequences entirely within time range
        TimeCriterion(
            start_after=datetime(2024, 6, 1),
            end_before=datetime(2024, 6, 30),
            sequence_within=True
        )

        # Combine both strict modes
        TimeCriterion(
            start_after=datetime(2024, 1, 1),
            end_before=datetime(2024, 12, 31),
            duration_within=True,
            sequence_within=True
        )
    """

    start_before: Optional[Union[datetime, str, int]] = None
    start_after: Optional[Union[datetime, str, int]] = None
    end_before: Optional[Union[datetime, str, int]] = None
    end_after: Optional[Union[datetime, str, int]] = None
    duration_within: bool = False
    sequence_within: bool = False
    date_format: Optional[str] = None
    # TODO: add settings for duration

    _REGISTER_NAME = "time"
