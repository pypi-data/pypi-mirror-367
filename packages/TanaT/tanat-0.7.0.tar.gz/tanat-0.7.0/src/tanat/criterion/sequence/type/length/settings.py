#!/usr/bin/env python3
"""
Length criterion.
"""

from typing import Optional

from pydantic.dataclasses import dataclass
from pypassist.dataclass.decorators.viewer.decorator import viewer

from ....base.settings import Criterion
from ....decorator.sequence import sequence_compatible
from .exception import ContradictoryLengthCriterionError


@viewer
@sequence_compatible
@dataclass
class LengthCriterion(Criterion):
    """
    Filter sequences based on their length.

    Provides flexible length-based filtering with multiple comparison
    operators for sequence selection.

    Args:
        gt (int, optional):
            Keep sequences with length strictly greater than this value.
        ge (int, optional):
            Keep sequences with length greater than or equal to this value.
        lt (int, optional):
            Keep sequences with length strictly less than this value.
        le (int, optional):
            Keep sequences with length less than or equal to this value.

    Example:
        # Filter sequences with 10 or fewer events
        LengthCriterion(le=10)
    """

    gt: Optional[int] = None  # greater than
    ge: Optional[int] = None  # greater than or equal to
    lt: Optional[int] = None  # less than
    le: Optional[int] = None  # less than or equal to

    _REGISTER_NAME = "length"

    def __post_init__(self):
        if self.gt is not None and self.ge is not None:
            raise ContradictoryLengthCriterionError(
                "Cannot use 'gt' and 'ge' operators simultaneously"
            )
        if self.lt is not None and self.le is not None:
            raise ContradictoryLengthCriterionError(
                "Cannot use 'lt' and 'le' operators simultaneously"
            )

        lower = self.gt if self.gt is not None else self.ge
        upper = self.lt if self.lt is not None else self.le
        if lower is not None and upper is not None and lower >= upper:
            raise ContradictoryLengthCriterionError(
                f"lower bound ({lower}) must be strictly less than upper bound ({upper})."
            )
