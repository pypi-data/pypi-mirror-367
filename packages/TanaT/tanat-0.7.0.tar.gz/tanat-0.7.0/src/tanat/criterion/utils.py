#!/usr/bin/env python3
"""
Utilities for criterion validation and resolution.
"""

from pypassist.mixin.registrable.exceptions import UnregisteredTypeError

from .base.enum import CriterionLevel
from .base.exception import InvalidCriterionError
from .base.settings import Criterion


def validate_criterion_params(criterion, level=None, criterion_type=None):
    """
    Validate criterion parameters and extract criterion_type if needed.

    Args:
        criterion (Union[Criterion, dict]): The criterion.
        level (str or CriterionLevel, optional): The criterion level.
        criterion_type (str, optional): The criterion type.

    Returns:
        str: The validated or extracted criterion_type.

    Raises:
        InvalidCriterionError: If validation fails.
    """
    if isinstance(criterion, dict):
        if criterion_type is None:
            raise InvalidCriterionError(
                "criterion_type must be provided if criterion is a dictionary."
            )
        if level is None:
            raise InvalidCriterionError(
                "level must be provided if criterion is a dictionary."
            )
    elif isinstance(criterion, Criterion):
        if criterion_type is None:
            # Extract criterion_type from criterion if not provided
            criterion_type = extract_criterion_type(criterion)
    elif not isinstance(criterion, Criterion) and not isinstance(criterion, dict):
        raise InvalidCriterionError(
            f"criterion must be a Criterion object or a dictionary. Got: {type(criterion)}"
        )

    return criterion_type


def resolve_criterion_level(criterion, provided_level=None, max_level=None):
    """
    Unified function to determine the correct CriterionLevel for any context.

    Args:
        criterion (Union[Criterion, dict]): The criterion.
        provided_level (str or CriterionLevel, optional): The provided criterion level.
        max_level (CriterionLevel, optional): The maximum level to consider.
            If None, defaults to SEQUENCE level for backward compatibility.

    Returns:
        CriterionLevel: The resolved criterion level.

    Raises:
        ValueError: If level must be provided but isn't (for trajectory context).
        InvalidCriterionError: If the provided level is incompatible with the criterion.
    """
    # Default compatibility levels based on max_level
    if max_level is None:
        # Backward compatibility: default to sequence context
        compatibility_levels = [CriterionLevel.SEQUENCE, CriterionLevel.ENTITY]
    else:
        # Full hierarchy up to max_level
        all_levels = [
            CriterionLevel.ENTITY,
            CriterionLevel.SEQUENCE,
            CriterionLevel.TRAJECTORY,
        ]
        compatibility_levels = [level for level in all_levels if level <= max_level]

    # Override with criterion-specific compatibility if available
    if isinstance(criterion, Criterion):
        compatibility_levels = criterion.get_compatibility_levels(max_level=max_level)

    # Check if level must be provided
    if len(compatibility_levels) > 1 and provided_level is None:
        # Use ValueError for trajectory context (backward compatibility)
        if max_level == CriterionLevel.TRAJECTORY:
            raise ValueError(
                "level must be provided if criterion is applicable at multiple levels."
            )

        raise InvalidCriterionError(
            "level must be provided if criterion is applicable at multiple levels."
        )

    # Convert string to enum if needed
    if isinstance(provided_level, str):
        provided_level = CriterionLevel.from_str(provided_level)

    # Validate compatibility
    if provided_level is not None and provided_level not in compatibility_levels:
        raise InvalidCriterionError(
            f"Level '{provided_level}' is not compatible with {criterion.__class__.__name__}. "
            f"Compatible levels: {compatibility_levels}"
        )

    # Return resolved level
    if provided_level is None:
        provided_level = compatibility_levels[0]

    return provided_level


# Backward compatibility alias
def resolve_trajectory_level(criterion, provided_level):
    """
    Backward compatibility wrapper for trajectory operations.

    Args:
        criterion (Union[Criterion, dict]): The criterion.
        provided_level (str or CriterionLevel, optional): The provided criterion level.

    Returns:
        CriterionLevel: The resolved criterion level.
    """
    return resolve_criterion_level(
        criterion, provided_level, max_level=CriterionLevel.TRAJECTORY
    )


def resolve_and_init_criterion(
    criterion, level=None, criterion_type=None, max_level=None
):
    """
    Ultimate function that validates, resolves level, and initializes criterion in one call.

    This function combines validate_criterion_params, resolve_criterion_level,
    and criterion initialization to eliminate code duplication across all methods.

    Args:
        criterion (Union[Criterion, dict]): The criterion to process.
        level (str or CriterionLevel, optional): The criterion level.
        criterion_type (str, optional): The criterion type.
        max_level (CriterionLevel, optional): The maximum level to consider.
            If None, defaults to SEQUENCE level context.

    Returns:
        tuple: (criterion_instance, resolved_level)
            - criterion_instance: Initialized criterion object ready to use
            - resolved_level: The resolved CriterionLevel

    Raises:
        InvalidCriterionError: If validation or resolution fails.
    """
    # Step 1: Validate and extract criterion_type
    criterion_type = validate_criterion_params(criterion, level, criterion_type)

    # Step 2: Resolve the appropriate level using unified function
    resolved_level = resolve_criterion_level(criterion, level, max_level)

    # Step 3: Get the registry class and initialize criterion
    criterion_base_cls = resolved_level.get_registry_cls()

    try:
        criterion_instance = criterion_base_cls.init(criterion, criterion_type)
    except UnregisteredTypeError as err:
        raise InvalidCriterionError(
            f"Criterion provided is not compatible for {resolved_level}"
        ) from err

    return criterion_instance, resolved_level


def extract_criterion_type(criterion):
    """
    Extract criterion type from criterion object.

    Args:
        criterion (Criterion): The criterion object.

    Returns:
        str: The criterion type.

    Raises:
        ValueError: If criterion_type cannot be extracted.
    """
    criterion_type = getattr(criterion, "_REGISTER_NAME", None)
    if criterion_type is None:
        raise ValueError(
            "Invalid settings: missing '_REGISTER_NAME' attribute. "
            "Use a valid settings class or specify criterion_type."
        )
    return criterion_type
