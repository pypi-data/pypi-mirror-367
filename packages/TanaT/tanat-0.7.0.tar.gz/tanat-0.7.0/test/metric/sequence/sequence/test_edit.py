#!/usr/bin/env python3
"""Test LCSSequenceMetric call method."""

import pytest

from tanat.metric.entity.type.hamming.metric import HammingEntityMetric
from tanat.metric.sequence.type.edit.metric import (
    EditSequenceMetric,
)
from tanat.metric.sequence.type.edit.settings import (
    EditSequenceMetricSettings,
)

from ...utils import replace_nan_with_value


class TestEditCall:
    """
    Test EditSequenceMetric call method.
    """

    @pytest.mark.parametrize("pool_type", ["event", "state", "interval"])
    def test_call_equivalence_with_snapshot(self, sequence_pools, pool_type, snapshot):
        """
        Ensure call methode gives the same result whether entity metric is passed as object or string.
        Snapshot is used to check the actual content.
        """
        pool = sequence_pools[pool_type]
        # Initialisation with HammingEntityMetric object
        metric_obj = EditSequenceMetric(
            settings=EditSequenceMetricSettings(entity_metric=HammingEntityMetric())
        )
        # Initialisation with string "hamming"
        metric_str = EditSequenceMetric(
            settings=EditSequenceMetricSettings(entity_metric="hamming")
        )

        ## -- sequence to compare
        seq_a = pool[2]
        seq_b = pool[3]

        # Call with SequenceMetric object
        value_obj = metric_obj(seq_a, seq_b)
        # Call with string
        value_str = metric_str(seq_a, seq_b)
        # Check consistency
        assert value_obj == value_str
        snapshot.assert_match(value_obj)
