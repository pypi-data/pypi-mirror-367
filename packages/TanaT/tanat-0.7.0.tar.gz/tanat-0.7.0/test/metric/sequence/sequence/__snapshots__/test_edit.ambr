# serializer version: 1
# name: TestEditCall.test_call_equivalence_with_snapshot[event]
  14.0
# ---
# name: TestEditCall.test_call_equivalence_with_snapshot[interval]
  6.0
# ---
# name: TestEditCall.test_call_equivalence_with_snapshot[state]
  3.0
# ---
